Test optimization and Wait API Tuning

slu-it opened the pull request:
I added some more tests and a rather big refactoring of the Wait API.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/testit-webtester/webtester2-core/11)

<!-- Reviewable:end -->

slu-it committed:
Refactored tests of browser operations to be truly isolated from outside influences

slu-it committed:
Added getter to conditions and created conditions factory test

slu-it committed:
Added unit tests for PageSourceSaver

slu-it committed:
Moved base configuration builder test

slu-it committed:
Added test for conditions

slu-it committed:
Isolated PostConstructInvokerTest

slu-it committed:
Added ScreenshotTaker unit test

slu-it committed:
Refactored the Wait API's internal class structure and overall design

The API (Wait.java) didn't actually change but  a lot of the internal classes did:
- Sleeper abstraction to allow for better testability
- Reduced visibility of most classes to the package
- Added additional wait operation for waiting until a given boolean supplier returns true
- Replaced WaitOperations with a much more streamlined and less complex class Waiter

slu-it commented on webtester-core/src/main/java/info/novatec/testit/webtester/waiting/Wait.java:
```diff
@@ -18,60 +17,87 @@
+     * The {@link Waiter} to use when executing wait operations.
+     * This field is package-private in order to allow for it to be tests.
+     */
+    static Waiter waiter = new Waiter();
```
I'm just 99% ok with this myself

slu-it commented on webtester-core/src/test/java/info/novatec/testit/webtester/conditions/ConditionTest.java:
```diff
@@ -0,0 +1,136 @@
+
+    }
+
+    public static class Or extends AbstractConditionTest {
```
Note to self: Names of Or tests are mixed up!

slu-it committed:
Added configuration tests

slu-it committed:
Moved static factory methods to end of PageFragmentProxyHandler class

slu-it committed:
Added unit tests for AdHoc Finder API

slu-it committed:
Cleanup JUnit Support by removing unused code and got rid of static class

slu-it committed:
Added Spring 4 support module tests

bitterblue commented:
Reviewed 35 of 43 files at r1, 11 of 11 files at r2, 13 of 13 files at r3, 6 of 6 files at r4, 6 of 6 files at r5, 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

_[webtester-core/src/main/java/info/novatec/testit/webtester/waiting/Wait.java, line 29 [r1]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/11#-KKtKCcf7zlhCX7fYvnW-r1-29:-KL6DaxeKeY388iXWmil:646634518) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/ddb604cd990c1cda498a478d754cce3b71741d52/webtester-core/src/main/java/info/novatec/testit/webtester/waiting/Wait.java#L29)):_
<details><summary>

<i>Previously, slu-it (Stefan Ludwig) wrote…</i></summary>



> I'm just 99% ok with this myself
> </details>
> The static scope is awkward, but it should work even for multiple threads, since WaitConfig is freshly instantiated and passed on for every call to the Wait API.

The package-private comment is no longer correct, is it?

---

_[webtester-core/src/test/java/info/novatec/testit/webtester/waiting/CurrentThreadSleeperTest.java, line 8 [r6]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/11#-KL6ceHQKqN7uibs6Ru_:-KL6ceHRNrOfNSynk4g6:-855585861) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/6444554680d17b0aecafeb72d7cfd76600db5955/webtester-core/src/test/java/info/novatec/testit/webtester/waiting/CurrentThreadSleeperTest.java#L8)):_

> ``` Java
> 
> 
> public class CurrentThreadSleeperTest {
> ```

I'm not happy with this test. `CurrentThreadSleeper` only encapsulates the static call to `currentThread.sleep()` so it can be mocked. I think, testing it creates more problems than the extra coverage would prevent.

---

_[webtester-core/src/test/java/info/novatec/testit/webtester/waiting/CurrentThreadSleeperTest.java, line 17 [r6]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/11#-KL6sK3gMNLL_R0_CCvi:-KL6sK3hKn2aTKgj3suy:-1027886693) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/6444554680d17b0aecafeb72d7cfd76600db5955/webtester-core/src/test/java/info/novatec/testit/webtester/waiting/CurrentThreadSleeperTest.java#L17)):_

> ``` Java
>         cut.sleep(100L);
>         long stop = now();
>         assertThat(stop - start).isGreaterThan(85).isLessThan(115);
> ```

This creates a race condition. The thread is not guaranteed to be resumed immediately after 100ms, it will only be rescheduled. Depending on OS and system load, the deviation can be a lot more than 15 ms.

---

_[webtester-core/src/test/java/info/novatec/testit/webtester/waiting/CurrentThreadSleeperTest.java, line 21 [r6]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/11#-KL6s_CYdfsK7QSPu88r:-KL6s_CYdfsK7QSPu88s:2114159256) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/6444554680d17b0aecafeb72d7cfd76600db5955/webtester-core/src/test/java/info/novatec/testit/webtester/waiting/CurrentThreadSleeperTest.java#L21)):_

> ``` Java
> 
>     private long now() {
>         return System.currentTimeMillis();
> ```

Uses system clock, so it is not a properly isolated unit test, more of an integration test against the clock. 

---

_[webtester-support-junit4/src/main/java/info/novatec/testit/webtester/junit/utils/ReflectionUtils.java, line 13 [r6]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/11#-KL6n3XvLbO3mnEb8rT-:-KL6n3XwPzCHnUtRy8HK:489385973) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/6444554680d17b0aecafeb72d7cfd76600db5955/webtester-support-junit4/src/main/java/info/novatec/testit/webtester/junit/utils/ReflectionUtils.java#L13)):_

> ``` Java
> public class ReflectionUtils {
> 
>     private final Map<Class<?>, Set<Field>> fieldsCache = new ConcurrentHashMap<>();
> ```

Is this cache still useful? It looks like the class is always instantiated freshly for single use.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/11)_

<!-- Sent from Reviewable.io -->

slu-it commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

_[webtester-core/src/main/java/info/novatec/testit/webtester/waiting/Wait.java, line 29 [r1]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/11#-KKtKCcf7zlhCX7fYvnW-r1-29:-KLBOgCzXVrO515d2eSQ:-1498987848) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/ddb604cd990c1cda498a478d754cce3b71741d52/webtester-core/src/main/java/info/novatec/testit/webtester/waiting/Wait.java#L29)):_
<details><summary>

<i>Previously, bitterblue (Axel Schüssler) wrote…</i></summary>



> The static scope is awkward, but it should work even for multiple threads, since WaitConfig is freshly instantiated and passed on for every call to the Wait API.
> 
> The package-private comment is no longer correct, is it?
> </details>
> Yeah, replaced that with a call to TestUtils.setStaticFieldValue(..)

---

_[webtester-core/src/test/java/info/novatec/testit/webtester/waiting/CurrentThreadSleeperTest.java, line 8 [r6]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/11#-KL6ceHQKqN7uibs6Ru_:-KLBROWwqrhga8MKBlju:2125846432) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/6444554680d17b0aecafeb72d7cfd76600db5955/webtester-core/src/test/java/info/novatec/testit/webtester/waiting/CurrentThreadSleeperTest.java#L8)):_
<details><summary>

<i>Previously, bitterblue (Axel Schüssler) wrote…</i></summary>



> I'm not happy with this test. `CurrentThreadSleeper` only encapsulates the static call to `currentThread.sleep()` so it can be mocked. I think, testing it creates more problems than the extra coverage would prevent.
> </details>
> hm.. I'm 50/50 on that one.  You're right that this is not really a unti test. Since it can't be isolated. On the other hand I like having a test for it in the unit testing "phase" as an early warning.
> But since it's very doubtful that anybody will accidentally change it in a breaking way AND it will still by tested implicitly with Wait\* integration tests, I'll remove it.

---

_[webtester-support-junit4/src/main/java/info/novatec/testit/webtester/junit/utils/ReflectionUtils.java, line 13 [r6]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/11#-KL6n3XvLbO3mnEb8rT-:-KLBPibPmXEr61Ws1luG:-2036350976) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/6444554680d17b0aecafeb72d7cfd76600db5955/webtester-support-junit4/src/main/java/info/novatec/testit/webtester/junit/utils/ReflectionUtils.java#L13)):_
<details><summary>

<i>Previously, bitterblue (Axel Schüssler) wrote…</i></summary>



> Is this cache still useful? It looks like the class is always instantiated freshly for single use.
> </details>
> You're right - since with the latestes changes a new instance is created with each JUnit Runner initialization the cache will never get used. I'll remove it.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/11)_

<!-- Sent from Reviewable.io -->

slu-it commented:
@bitterblue changed commits according to feedback.

bitterblue commented:
Reviewed 1 of 2 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

_[webtester-core/src/main/java/info/novatec/testit/webtester/waiting/Wait.java, line 29 [r1]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/11#-KKtKCcf7zlhCX7fYvnW-r1-29:-KLBj8Rge0HBMT9Jd1js:-1840120511) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/ddb604cd990c1cda498a478d754cce3b71741d52/webtester-core/src/main/java/info/novatec/testit/webtester/waiting/Wait.java#L29)):_
<details><summary>

<i>Previously, slu-it (Stefan Ludwig) wrote…</i></summary>



> Yeah, replaced that with a call to TestUtils.setStaticFieldValue(..)
> </details>
> You forgot to remove it, then ;-)

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/11)_

<!-- Sent from Reviewable.io -->

slu-it committed:
Changed Wait's Waiter field from a static reference to a supplier

This allows for testability as well as changing the behavior of the Wait class with custom behavior.
The default behavior is thread safe because each call gets a new instance of Waiter.

slu-it commented:
Review status: 68 of 73 files reviewed at latest revision, 1 unresolved discussion.

---

_[webtester-core/src/main/java/info/novatec/testit/webtester/waiting/Wait.java, line 29 [r1]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/11#-KKtKCcf7zlhCX7fYvnW-r1-29:-KLC6ivsvGjti0tub0Im:-1825244729) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/ddb604cd990c1cda498a478d754cce3b71741d52/webtester-core/src/main/java/info/novatec/testit/webtester/waiting/Wait.java#L29)):_
<details><summary>

<i>Previously, bitterblue (Axel Schüssler) wrote…</i></summary>



> You forgot to remove it, then ;-)
> </details>
> hm.. in the latest version the field should be private - but i had an idea how to make this more flexible as well as thread safe. Take a look at the newes commit #7489085

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/11)_

<!-- Sent from Reviewable.io -->

bitterblue commented:
Review status: 68 of 73 files reviewed at latest revision, 1 unresolved discussion.

---

_[webtester-core/src/main/java/info/novatec/testit/webtester/waiting/Wait.java, line 29 [r1]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/11#-KKtKCcf7zlhCX7fYvnW-r1-29:-KLCGxW-LlTHy-Daf1tl:-1017490848) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/ddb604cd990c1cda498a478d754cce3b71741d52/webtester-core/src/main/java/info/novatec/testit/webtester/waiting/Wait.java#L29)):_
<details><summary>

<i>Previously, slu-it (Stefan Ludwig) wrote…</i></summary>



> hm.. in the latest version the field should be private - but i had an idea how to make this more flexible as well as thread safe. Take a look at the newes commit #7489085
> </details>
> way more flexible. Nice :)

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/11)_

<!-- Sent from Reviewable.io -->

bitterblue commented:
Reviewed 5 of 5 files at r8.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/11)_

<!-- Sent from Reviewable.io -->

slu-it commented:
Reviewed 1 of 5 files at r8.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/11)_

<!-- Sent from Reviewable.io -->

slu-it commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

_[webtester-core/src/main/java/info/novatec/testit/webtester/waiting/Wait.java, line 29 [r1]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/11#-KKtKCcf7zlhCX7fYvnW-r1-29:-KLCYXWra-b3R-rJfEDE:-499146009) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/ddb604cd990c1cda498a478d754cce3b71741d52/webtester-core/src/main/java/info/novatec/testit/webtester/waiting/Wait.java#L29)):_
<details><summary>

<i>Previously, bitterblue (Axel Schüssler) wrote…</i></summary>



> way more flexible. Nice :)
> </details>
> Done.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/11)_

<!-- Sent from Reviewable.io -->

slu-it commented:
I'm a bit confused by the open reviewable state. @bitterblue are you done?
Travis seams to be having problems as well - but it did build the branch on my sandbox....

bitterblue commented:
To me it looks like you are still having a discussion with codacy-bot on reviewable ... (-_-
I also see travis hanging. Strange.

slu-it commented:
Reviewed 35 of 43 files at r1, 5 of 11 files at r2, 13 of 13 files at r3, 6 of 6 files at r4, 5 of 6 files at r5, 4 of 4 files at r6, 1 of 2 files at r7, 4 of 5 files at r8.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/11)_

<!-- Sent from Reviewable.io -->

slu-it commented:
hm.. lesson: don't reply to the bot - if it does not acknowledge the review remains open - good that i'm admin :D
